It seems stateful test cached some bug

```
E       TypeError: '<' not supported between instances of 'NoneType' and 'str'
E       Falsifying example:
E       state = ModifiedZarrHierarchyStateMachine()
E       state.init_store()
E       state.check_list_prefix_from_root()
E       Checking 1 expected keys vs 1 actual keys
E       Draw 1: bool_(False)
E       Adding array:  path='0'  shape=(1,)  chunks=(1,)
E       state.add_array(array_and_chunks=(array([False]), (1,)), data=data(...), name='0')
E       state.check_list_prefix_from_root()
E       Checking 2 expected keys vs 2 actual keys
E       Draw 2: bool_(False)
E       Adding array:  path='-'  shape=(2,)  chunks=(2,)
E       state.add_array(array_and_chunks=(array([ True, False]), (2,)), data=data(...), name='-')
E       state.check_list_prefix_from_root()
E       Checking 4 expected keys vs 4 actual keys
E       committing and checking list_prefix
E       Draw 3: '-/c/0'
E       state.commit_with_check(data=data(...))
E       state.check_list_prefix_from_root()
E       Checking 4 expected keys vs 4 actual keys
E       Draw 4: '-'
E       Draw 5: '0'
E       deleting chunk path='-/c/0'
E       state.delete_chunk(data=data(...))
E       state.check_list_prefix_from_root()
E       Checking 3 expected keys vs 3 actual keys
E       committing and checking list_prefix
E       Draw 6: '-/zarr.json'
E       state.commit_with_check(data=data(...))
E       state.teardown()
```

--- a/icechunk-python/tests/test_zarr/test_store/test_icechunk_store.py
+++ b/icechunk-python/tests/test_zarr/test_store/test_icechunk_store.py
@@ -2,13 +2,13 @@ from __future__ import annotations
 
 import pickle
 from pathlib import Path
-from typing import Any
+from typing import Any, TypeVar
 
 import pytest
 
 from icechunk import IcechunkStore, local_filesystem_storage
 from icechunk.repository import Repository
-from zarr.abc.store import OffsetByteRequest, RangeByteRequest, SuffixByteRequest
+from zarr.abc.store import OffsetByteRequest, RangeByteRequest, Store, SuffixByteRequest
 from zarr.core.buffer import Buffer, cpu, default_buffer_prototype
 from zarr.core.sync import collect_aiterator
 from zarr.testing.store import StoreTests
@@ -20,6 +20,8 @@ ARRAY_METADATA = (
     b'"chunk_key_encoding":{"name":"default","configuration":{"separator":"/"}},"fill_value":0,'
     b'"codecs":[{"name":"mycodec","configuration":{"foo":42}}],"storage_transformers":[{"name":"mytransformer","configuration":{"bar":43}}],"dimension_names":["x","y","t"]}'
 )
+
+S = TypeVar("S", bound=Store)
 
 
 class TestIcechunkStore(StoreTests[IcechunkStore, cpu.Buffer]):
@@ -152,9 +154,6 @@ class TestIcechunkStore(StoreTests[IcechunkStore, cpu.Buffer]):
     def test_store_supports_partial_writes(self, store: IcechunkStore) -> None:
         assert not store.supports_partial_writes
 
-    async def test_list_prefix(self, store: IcechunkStore) -> None:
-        assert True
-
     async def test_clear(self, store: IcechunkStore) -> None:
         await self.set(
             store,
@@ -186,6 +185,27 @@ class TestIcechunkStore(StoreTests[IcechunkStore, cpu.Buffer]):
         )
         keys = [k async for k in store.list()]
         assert keys == ["foo/zarr.json"], keys
+
+    async def test_list_prefix(self, store: S) -> None:
+        """
+        Test that the `list_prefix` method works as intended. Given a prefix, it should return
+        all the keys in storage that start with this prefix.
+        """
+        prefixes = ("", "a/", "a/b/", "a/b/c/")
+        data = self.buffer_cls.from_bytes(DEFAULT_GROUP_METADATA)
+        fname = "zarr.json"
+        store_dict = {p + fname: data for p in prefixes}
+
+        await store._set_many(store_dict.items())
+
+        for prefix in prefixes:
+            observed = tuple(sorted([_ async for _ in store.list_prefix(prefix)]))
+            expected: tuple[str, ...] = ()
+            for key in store_dict:
+                if key.startswith(prefix):
+                    expected += (key,)
+            expected = tuple(sorted(expected))
+            assert observed == expected
 
     async def test_list_dir(self, store: IcechunkStore) -> None:
         out = [k async for k in store.list_dir("")]

--- a/icechunk/src/session.rs
+++ b/icechunk/src/session.rs
@@ -1258,6 +1258,13 @@ impl<'a> FlushProcess<'a> {
                         )
                         .clone()
                 }));
+                for mr in array_refs.iter() {
+                    let new_ref = mr.clone();
+                    self.manifest_refs
+                        .entry(node.id.clone())
+                        .and_modify(|v| v.push(new_ref.clone()))
+                        .or_insert_with(|| vec![new_ref]);
+                }
             }
             NodeData::Group => {}
         }
@@ -1311,17 +1318,13 @@ async fn flush(
     .map(|node| {
         let id = &node.id;
         // TODO: many clones
-        if let NodeData::Array(meta, original_manifests) = node.node_data {
-            if let Some(manifests) = flush_data.manifest_refs.get(id) {
-                NodeSnapshot {
-                    node_data: NodeData::Array(meta.clone(), manifests.clone()),
-                    ..node
-                }
-            } else {
-                NodeSnapshot {
-                    node_data: NodeData::Array(meta, original_manifests),
-                    ..node
-                }
+        if let NodeData::Array(meta, _) = node.node_data {
+            NodeSnapshot {
+                node_data: NodeData::Array(
+                    meta.clone(),
+                    flush_data.manifest_refs.get(id).cloned().unwrap_or_default(),
+                ),
+                ..node
             }
         } else {
             node

--- a/icechunk/src/session.rs
+++ b/icechunk/src/session.rs
@@ -2462,6 +2462,38 @@ mod tests {
         let size_after_chunk_delete = manifest.len();
         assert!(size_after_chunk_delete < size_after_delete);
 
+        // delete the second chunk, now there are no chunks, so there should be no manifests either
+        let mut ds = repo.writable_session("main").await?;
+        ds.set_chunk_ref(a1path.clone(), ChunkIndices(vec![0, 1]), None).await?;
+        let _snap_id = ds.commit("chunk deleted", None).await?;
+
+        let manifests = match ds.get_array(&a1path).await?.node_data {
+            NodeData::Array(_, manifests) => manifests,
+            NodeData::Group => panic!("must be an array"),
+        };
+        assert!(manifests.is_empty());
+
+        // there should be three manifests (unchanged)
+        assert_eq!(
+            3,
+            in_mem_storage
+                .all_keys()
+                .await?
+                .iter()
+                .filter(|key| key.contains("manifest"))
+                .count()
+        );
+        // there should be six snapshots
+        assert_eq!(
+            6,
+            in_mem_storage
+                .all_keys()
+                .await?
+                .iter()
+                .filter(|key| key.contains("snapshot"))
+                .count(),
+        );
+
         Ok(())
     }
 

